Riot/Modules/Authentication/AuthenticationViewController.m

```diff
@@ -1146,26 +1146,82 @@ - (void)sessionStateDidChangeNotification:(NSNotification*)notification
+                                NSLog(@"[AuthenticationVC] sessionStateDidChange: Bootstrap succeeded");
+                                [self dismiss];
+                            } failure:^(NSError * _Nonnull error) {
+                                NSLog(@"[AuthenticationVC] sessionStateDidChange: Bootstrap failed. Error: %@", error);
+                                [session.crypto setOutgoingKeyRequestsEnabled:YES onComplete:nil];
+                                [self dismiss];
+                            }];
+                        }
+                        else
+                        {
+                            NSLog(@"[AuthenticationVC] sessionStateDidChange: Do not know how to bootstrap");
+                            
+                            [session.crypto setOutgoingKeyRequestsEnabled:YES onComplete:nil];
+                            [self dismiss];
+                        }
+#else
+                        [session.crypto setOutgoingKeyRequestsEnabled:YES onComplete:nil];
+                        [self dismiss];
+#endif
+                        break;
+                    }
+                    case MXCrossSigningStateCrossSigningExists:
+                    {
+                        NSLog(@"[AuthenticationVC] sessionStateDidChange: Complete security");
                         
-                        self.userInteractionEnabled = YES;
-                        [self.authenticationActivityIndicator stopAnimating];
+                        // Ask the user to verify this session
+                        dispatch_async(dispatch_get_main_queue(), ^{
+                            
+                            self.userInteractionEnabled = YES;
+                            [self.authenticationActivityIndicator stopAnimating];
+                            
+                            [self presentCompleteSecurityWithSession:session];
+                        });
+                        break;
+                    }
                         
-                        [self presentCompleteSecurityWithSession:session];
-                    });
-                }
-                else
-                {
-                    [session.crypto setOutgoingKeyRequestsEnabled:YES onComplete:nil];
-                    [self dismiss];
+                    default:
+                        NSLog(@"[AuthenticationVC] sessionStateDidChange: Nothing to do");
+                        
+                        [session.crypto setOutgoingKeyRequestsEnabled:YES onComplete:nil];
+                        [self dismiss];
+                        break;
                 }
                 
             } failure:^(NSError * _Nonnull error) {
-                NSLog(@"[AuthenticationVC] Fail to refresh crypto state with error: %@", error);
+                NSLog(@"[AuthenticationVC] sessionStateDidChange: Fail to refresh crypto state with error: %@", error);
                 dispatch_async(dispatch_get_main_queue(), ^{
```

Comment:
Callbacks are called on main thread we can remove `dispatch_async(dispatch_get_main_queue(), ^{` calls.